h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -2126,6 +2134,160 @@ private void fitIRLSM(Solver s) {
+          }
+
+          // Dispersion estimation part
+          if (negativebinomial.equals(_parms._family)){
+            converged = updateNegativeBinomialDispersion(iterCnt, _state.beta(), previousLLH, weights, response) && converged;
+          }
+          if (Math.abs(previousLLH - gram.likelihood) < _parms._objective_epsilon)
+            sameLLH ++;
+          else
+            sameLLH = 0;
+          converged = converged || sameLLH > 10;
+          previousLLH = gram.likelihood;
+        }
+      } catch (NonSPDMatrixException e) {
+        Log.warn(LogMsg("Got Non SPD matrix, stopped."));
+        throw e; // TODO: Decide if we want to return a bad model or no model at all.
+      }
+    }
+
+    private void updateTheta(double theta){
+      if (_state._glmw != null) {
+         _state._glmw._theta = theta;
+         _state._glmw._invTheta = 1./theta;
+      }
+      _parms._theta = theta;
+      _parms._invTheta =1./theta;
+      _model._parms._theta = theta;
+      _model._parms._invTheta = 1./theta;
+    }
+
+    private boolean updateNegativeBinomialDispersion(int iterCnt, double[] betaCnd, double previousNLLH, Vec weights, Vec response) {
+      double delta;
+      double theta;
+      boolean converged = false;
+      if (iterCnt == 1) {
+        theta = estimateNegBinomialDispersionMomentMethod(_model, betaCnd, _dinfo, weights, response);
+      } else {
+        theta = _parms._theta;
+        DispersionTask.GenPrediction gPred = new DispersionTask.GenPrediction(betaCnd, _model, _dinfo).doAll(
+                1, Vec.T_NUM, _dinfo._adaptedFrame);
+        Vec mu = gPred.outputFrame(Key.make(), new String[]{"prediction"}, null).vec(0);
+
+        NegativeBinomialGradientAndHessian nbGrad = new NegativeBinomialGradientAndHessian(theta).doAll(mu, response, weights);
+        delta = _parms._dispersion_learning_rate * nbGrad._grad / nbGrad._hess;
+        double bestLLH = Math.max(-previousNLLH, nbGrad._llh);
+        double bestTheta = theta;
+
+        delta = Double.isFinite(delta) ? delta : 1; // NaN can occur in extreme datasets so try to get out of this neighborhood just by linesearch
+
+        // Golden section search for the optimal size of delta
+        // Set lowerbound to -10 or lowest value that will keep theta > 0 which ever is bigger
+        // Negative value here helps with datasets where we use to diverge, I'm not sure yet if it's caused by some
+        // numerical issues or if the likelihood can get multimodal for some cases.
+        double lowerBound = (theta + 10 * delta < 0) ? (1 - 1e-15) * theta / delta : -10;
+        double upperBound = (theta - 1e3 * delta < 0) ? (1 - 1e-15) * theta / delta : 1e3;
+        double d = upperBound - lowerBound;
+
+        for (int i = 0; i < _parms._max_iterations_dispersion; i++) {
+          d *= 0.618;
+          final double lowerBoundProposal = upperBound - d;
```

Comment:
The d is 1/golden ratio.  Show some respect and make a comment about it please.

Reply:
I added a comment about it there.